### PR TITLE
Improve RichProgreasBar speed performance 

### LINKED ISSUE #13937: `RichProgressBar` in v1.6 is slower than v1.5
## 🐛 Bug
Similarly to (but independently of) #13179, there's a regression in speed of the rich progress bar between 1.5 and 1.6.

```
time: 2.059369374997914   # v1.5.10
time: 13.186531708983239  # v1.6.0
time: 11.988152374979109  # v1.6.5
time: 12.897805292042904  # master (aefb9ab43f9a8e6704558a346dbae1a00044bb45)
```

### To Reproduce

```python
from time import monotonic
import torch
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import RichProgressBar

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]

def main():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        max_epochs=100,
        enable_model_summary=False,
        enable_checkpointing=False,
        logger=False,
        benchmark=False,  # True by default in 1.6.{0-3}.
        callbacks=RichProgressBar(),
    )
    t0 = monotonic()
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    print("time:", monotonic() - t0)

if __name__ == "__main__":
    main()
```

### Expected behavior
There should be no regression unless there's a reasonable explanation.

### Environment

- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): n/a
- PyTorch Lightning Version (e.g., 1.5.0): master (aefb9ab43f9a8e6704558a346dbae1a00044bb45), 1.6.5, 1.6.0, 1.5.10
- PyTorch Version (e.g., 1.10): 1.12
- Python version (e.g., 3.9): 3.10
- OS (e.g., Linux): macOS
- CUDA/cuDNN version: n/a
- GPU models and configuration: Apple Silicon M1
- How you installed PyTorch (`conda`, `pip`, source): pip
- If compiling from source, the output of `torch.__config__.show()`:
- Running environment of LightningApp (e.g. local, cloud): n/a
- Any other relevant information: `rich==12.5.1`

### Additional context
Investigating!

https://github.com/Lightning-AI/lightning/pulls?q=is%3Apr+label%3A%22progress+bar%3A+rich%22+milestone%3A%22pl%3A1.6%22+

cc @tchaton @rohitgr7 @kaushikb11 @borda @akihironitta

### PR DESCRIPTION
**Pull request by GdoongMathew**
## What does this PR do?

Attempt to fix #13937

Benchmark Results are as below

```text
master
    === Benchmark Results ===
    Version      No Rich (s)    With Rich (s)
    ---------  -------------  ---------------
    2.6.0.dev          7.945           13.089

master auto-refresh enabled, 33 updates / second
    === Benchmark Results ===
    Version      No Rich (s)    With Rich (s)
    ---------  -------------  ---------------
    2.6.0.dev          7.605             7.09

this branch
    === Benchmark Results ===
    Version      No Rich (s)    With Rich (s)
    ---------  -------------  ---------------
    2.6.0.dev          7.676            8.345
```

Key changes:
1.  Conditional refresh
    `CustomProgress.refresh` now only calls its superclass method when running in interactive mode. This change is critical for performance since in most cases, we only need to `notify` the refresh thread rather than performing a full refresh.
2. No `threading.Event` for custom `_RefreshThread`
  Using `Event` would introduce unnecessary locking overhead. Since acquiring and releasing locks are relatively expensive here, this approach didn’t yield any speed improvement.
3. Auto-refresh side effects
  Simply enabling `auto_refresh` would revert #10497 and bring back #10607, which causes pdb unusable when using `RichProgressBar`.

TBH, I’m not entirely sure if this is the best solution, but I haven’t found another way to improve performance without undoing prior fixes.
Given that we’re now defaulting to rich-based progress bars and model summaries when rich is installed, I think this could still be quite a beneficial improvement — happy to hear thoughts or alternative ideas.

### BC Change
* refresh_rate: The meaning of `refresh_rate` is changed from `update per iteration` to `update per second`.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21321.org.readthedocs.build/en/21321/

### CODE REVIEW

File: src/lightning/pytorch/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -66,9 +69,49 @@ class CustomInfiniteTask(Task):
...
+            self.refresh_cond = False
+            super().__init__(*args, **kwargs)
+
+        def run(self) -> None:
+            while not self.done.is_set():
+                if self.refresh_cond:
+                    with self.live._lock:
+                        self.live.refresh()
+                    self.refresh_cond = False
+                time.sleep(0.005)
```
Comment (justusschock) [inline src/lightning/pytorch/callbacks/progress/rich_progress.py]: Looks great so far!
One minor question: can we make this configurable? so we/users can easily reduce this even more if required?

Comment (GdoongMathew) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py]: What do you think about reusing the original [`_RefreshThread.refresh_per_second`](https://github.com/Textualize/rich/blob/master/rich/live.py#L27), then exposing it as a configurable argument in the `RichProgressBar` callback.

One thing to note, this PR would also nullify the current `refresh_rate` argument, since we are now auto-refreshing the progress with `refresh_per_second` frequency.

Comment (justusschock) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py]: couldn't we use the refresh_rate argument instead then? But yeah generally something like that I had in mind :)

Comment (GdoongMathew) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py]: Yes, that works. I believe this is also related to the discussion mentioned [here](https://github.com/Lightning-AI/pytorch-lightning/pull/9647#discussion_r736391507). I agree that we should keep using `refresh_rate` in this case, as long as we’re okay with updating its description, since it'll now represent the **updates per second** rather than current update frequency in **number of batches**?

Comment (GdoongMathew) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py]: Fixed

**Review by justusschock — APPROVED** [review]
